openmdao/core/group.py

```diff
@@ -1267,20 +1270,20 @@ def _solve_nonlinear(self):
 
         name = self.pathname if self.pathname else 'root'
 
-        # Execute guess_nonlinear if specified.
-        # We need to call this early enough so that any solver that needs initial guesses has
-        # them.
-        # TODO: It is pointless to run this ahead of non-iterative solvers.
-        for sub in self.system_iter(recurse=True):
-            if hasattr(sub, 'guess_nonlinear'):
-                with sub._unscaled_context(outputs=[sub._outputs], residuals=[sub._residuals]):
-                    sub.guess_nonlinear(sub._inputs, sub._outputs, sub._residuals)
-
         with Recording(name + '._solve_nonlinear', self.iter_count, self):
             result = self._nonlinear_solver.solve()
 
         return result
 
+    def _guess_nonlinear(self):
+        """
+        Provide initial guess for states.
+        """
+        for isub, sub in enumerate(self._subsystems_myproc):
+            if self._has_guess:
```

Comment:
I would put the if self._has_guess check outside of the loop, and then maybe put an 'if sub._has_guess' inside the loop.